programs/stake/src/stake_state.rs

```diff
@@ -1005,10 +1019,30 @@ impl MergeKind {
                     source_meta.rent_exempt_reserve,
                     source_stake.delegation.stake,
                 )?;
+                if invoke_context
+                    .is_feature_active(&stake_merge_with_unmatched_credits_observed::id())
+                {
+                    stake.credits_observed = stake_weighted_credits_observed(
+                        &stake,
+                        source_lamports,
+                        source_stake.credits_observed,
+                    )
+                    .ok_or(InstructionError::ArithmeticOverflow)?;
+                }
                 stake.delegation.stake = checked_add(stake.delegation.stake, source_lamports)?;
                 Some(StakeState::Stake(meta, stake))
             }
             (Self::FullyActive(meta, mut stake), Self::FullyActive(_, source_stake)) => {
+                if invoke_context
+                    .is_feature_active(&stake_merge_with_unmatched_credits_observed::id())
+                {
+                    stake.credits_observed = stake_weighted_credits_observed(
+                        &stake,
+                        source_stake.delegation.stake,
+                        source_stake.credits_observed,
+                    )
+                    .ok_or(InstructionError::ArithmeticOverflow)?;
+                }
```

Comment:
how about extracting the duplicate new code like this?

```rust
fn merge_delegation_stake_and_credits_observed(
    invoke_context: TTTT,
    stake: &Stake,
    merge_lamports: u64,
    merge_credits_observed: u64) {
    if invoke_context
        .is_feature_active(&stake_merge_with_unmatched_credits_observed::id())
    {
       stake.credits_observed = stake_weighted_credits_observed(
            &stake,
            merge_lamports,
            merge_credits_observed,
      )
      .ok_or(InstructionError::ArithmeticOverflow)?;
    }
    stake.delegation.stake =
      checked_add(merge_lamports, merge_credits_observed)?;
    Some(StakeState::Stake(meta, stake))
 }

...

            (
                Self::ActivationEpoch(meta, mut stake),
                Self::ActivationEpoch(source_meta, source_stake),
            ) => {
               // extracted common code like below
            },
            (Self::FullyActive(meta, mut stake), Self::FullyActive(_, source_stake)) => {
                // Don't stake the source account's `rent_exempt_reserve` to
                // protect against the magic activation loophole. It will
                // instead be moved into the destination account as extra,
                // withdrawable `lamports`
                merge_delegation_stake_and_credits_observed(....)
            }

```

Reply:
Great suggestion, will do